README.md
# Bulldog

<img src="./bulldog-logo.svg" alt="logo" width="64px" height="64px"/>

__Bulldog__ is an amazing hunting dog that checks for you a list of URLs and
warns you by email if one of them returns a http code that is not 200.

## Installation

Install it yourself

    $ go get github.com/pioz/bulldog
    $ cd $GOPATH/src/github.com/pioz/bulldog
    $ go build
    $ ./bulldog -v

or
[⬇️ download](https://github.com/pioz/bulldog/blob/master/deb/bulldog_0.2.1-1_amd64.deb?raw=true)
and install the `.deb` package with

    $ dpkg -i bulldog_0.2.1-1_amd64.deb

## Usage

Bulldog loads a list of URLs and for each of them executes a GET http request.
If the request fails or returns a different http code than 200, it will send you
an email to alert you. When the list is over it starts to sleep for a while and
then restarts the controls.

Bulldog uses a configuration file with various options:

```ini
[time]
# After checking the entire list of URLs sleep for these seconds.
s=60
# After checking the entire list of URLs and at least a check fail sleep for
# these seconds. Usually this time is greater to not warn you continuously.
se=600
# Http request timeout. If the timeout is reached the check is to be considered
# as failed.
t=10
[logging]
# Log file path. If empty logs on stdout.
logfile=/var/log/bulldog.log
# Disables logs.
quiet=false
[email]
# Gmail account. If this is present send email using the gmail smtp server. Use
# -pass flag to specify the gmail account password. If this flag is empty send
# email using `mail` command line program.
gmail=account@gmail.com
# Gmail account password. Only relevant when using -gmail flag.
pass=pa$$w0rd
# When a check fails send an email on this email address. If is empty the email
# alert is disabled.
to=your@email.com
[urls]
# Comma-separated list of URLs to check.
urls=http://google.com,http://twitter.com/
```

To unleash Bulldog run the follow command:

    $ bulldog -config /path/to/config/file

You can pass also the config file options as command arguments:

    $ bulldog -urls http://google.com -to your@email.com -s 10

You can also perform a check on the list only once and then exit (status code 1
if at least one check fails):

    $ bulldog -urls http://google.com,http://twitter.com -1

For the complete list of command line arguments:

    $ bulldog -h

### init.d script

[Instruction here](https://github.com/pioz/bulldog/wiki/init.d-script-for-Debian).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/pioz/bulldog.

## License

The package is available as open source under the terms of the [GPL License](https://github.com/pioz/bulldog/blob/master/LICENSE).

deb/changelog
bulldog (0.2.1-1) UNRELEASED; urgency=medium

  * If -1 flag is present exit with status code 1 if at least one check fails,
  otherwise exit with status code 0.

 -- Enrico Pilotto <epilotto@gmx.com>  Wed, 01 Nov 2017 13:59:37 +0100

bulldog (0.2.0-1) UNRELEASED; urgency=medium

  * Send one email with a report of all errors

  * Fix conf bugs

 -- Enrico Pilotto <epilotto@gmx.com>  Tue, 31 Oct 2017 18:03:24 +0100

bulldog (0.1.0-2) UNRELEASED; urgency=medium

  * Initial release (Closes: #2432)

 -- Enrico Pilotto <epilotto@gmx.com>  Tue, 31 Oct 2017 02:01:53 +0100


mailer.go
package main

import (
	"fmt"
	"log"
	"net/smtp"
	"net/url"
	"os/exec"
)

// Mailer struct
type Mailer struct {
	gmail, pass, to string
}

// BuildAndSendEmail send alert email
func (m *Mailer) BuildAndSendEmail(errors []error) {
	if m.to != "" {
		var (
			count   = len(errors)
			subject string
			body    string
		)
		if count == 1 {
			urlErr, ok := errors[0].(*url.Error)
			if ok {
				subject = fmt.Sprintf("'%s' is unreachable", urlErr.URL)
			} else {
				subject = "A site is unreachable"
			}
		} else {
			subject = fmt.Sprintf("%d sites are unreachable", count)
		}
		for _, error := range errors {
			body += fmt.Sprintf("* %s\n\n", error.Error())
		}
		err := m.SendMail(subject, body)
		if err != nil {
			log.Println(err)
		}
	}
}

// SendMail send a email
func (m *Mailer) SendMail(subject, body string) error {
	if m.gmail == "" {
		return m.sendEmailWithMail(subject, body)
	}
	return m.sendEmailWithGmail(subject, body)
}

func (m *Mailer) sendEmailWithGmail(subject, body string) error {
	msg := "From: Bulldog <bulldog>\n" +
		"To: " + m.to + "\n" +
		"Subject: [Bulldog] " + subject + "\n\n" +
		body
	err := smtp.SendMail("smtp.gmail.com:587",
		smtp.PlainAuth("", m.gmail, m.pass, "smtp.gmail.com"),
		m.gmail, []string{m.to}, []byte(msg))
	if err != nil {
		return err
	}
	return nil
}

func (m *Mailer) sendEmailWithMail(subject, body string) error {
	c1 := exec.Command("echo", body)
	c2 := exec.Command("mail", "-s [Bulldog] "+subject, "-r Bulldog <bulldog>", m.to)
	c2.Stdin, _ = c1.StdoutPipe()
	err1 := c1.Start()
	err2 := c2.Run()
	err3 := c1.Wait()
	if err1 != nil {
		return err1
	}
	if err2 != nil {
		return err2
	}
	if err3 != nil {
		return err3
	}
	return nil
}

// func errorToEmoji(err error) string {
// 	if strings.Contains(err.Error(), "Client.Timeout exceeded") {
// 		return "⌛️"
// 	}
// 	if strings.Contains(err.Error(), "unsupported protocol scheme") {
// 		return "😓"
// 	}
// 	if strings.Contains(err.Error(), "no such host") {
// 		return "🛑"
// 	}
// 	if strings.Contains(err.Error(), "status code is 403") {
// 		return "🔒"
// 	}
// 	if strings.Contains(err.Error(), "status code is 404") {
// 		return "🔎"
// 	}
// 	if strings.Contains(err.Error(), "status code is 500") {
// 		return "🆘"
// 	}
// 	return "⚠️"
// }


deb/makedeb.sh
#!/bin/bash

ARCHS="i386 amd64"
if ! echo $ARCHS | grep -w "$ARCH" > /dev/null ; then
  echo "You need to set a valid $ARCH variable (i386|amd64)"
  exit 2
fi

DEBEMAIL="epilotto@gmx.com"
DEBFULLNAME="Enrico Pilotto"
export DEBEMAIL DEBFULLNAME

NAME="bulldog"
VERSION=`cat VERSION`
PKGDIR=$NAME-$VERSION

rm -rf $PKGDIR
rm -rf ${NAME}_$VERSION*

mkdir -p $PKGDIR/usr/bin
cp ${NAME}_$ARCH $PKGDIR/usr/bin/$NAME
cp -r etc $PKGDIR
cd $PKGDIR
dh_make -s --createorig --copyright=gpl3 -y
cd ..
rm $PKGDIR/debian/README*
cp $NAME.1 $NAME.manpages changelog copyright install postinst postrm VERSION $PKGDIR/debian
cp control_$ARCH $PKGDIR/debian/control
cd $PKGDIR
dpkg-buildpackage -us -uc
cd ..
debc ${NAME}_$VERSION-1_$ARCH.changes
lintian ${NAME}_$VERSION-1_$ARCH.deb

rm -rf ${NAME}-$VERSION
rm ${NAME}_$VERSION-1.debian.tar.xz ${NAME}_$VERSION-1.dsc ${NAME}_$VERSION-1_amd64.buildinfo ${NAME}_$VERSION-1_amd64.changes ${NAME}_$VERSION.orig.tar.xz


bulldog.go
package main

import (
	"bufio"
	"errors"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"net/url"
	"os"
	"strings"
	"time"

	"github.com/vharitonsky/iniflags"
)

// Implement flag.Value interface
type urls []string

func (u *urls) String() string {
	return fmt.Sprint(*u)
}

func (u *urls) Set(value string) error {
	if len(*u) > 0 {
		return errors.New("urls flag already set")
	}
	for _, url := range strings.Split(value, ",") {
		if url != "" {
			*u = append(*u, url)
		}
	}
	return nil
}

// Config is a struct that contains config information
type Config struct {
	urlFile               string
	sleep, sleepWithError int
	timeout               int
	oneCheck              bool
	logFile               string
	quiet                 bool
	gmail, pass, to       string
	urls                  urls
}

func configure(config *Config) {
	flag.StringVar(&config.urlFile, "f", "", "File that contains the urls to check each per line.")

	flag.IntVar(&config.sleep, "s", 60, "Seconds to sleep between a complete URLs check to another.")
	flag.IntVar(&config.sleepWithError, "se", 600, "Seconds to sleep between a complete URLs check to another if a check has failed.")
	flag.IntVar(&config.timeout, "t", 10, "Check request timeout in seconds.")
	flag.BoolVar(&config.oneCheck, "1", false, "Perform URLs checks only one time then exit with status code 1 if at least one check fails, otherwise exit with code 0.")

	flag.StringVar(&config.logFile, "logfile", "", "Logfile path.")
	flag.BoolVar(&config.quiet, "quiet", false, "Disable logging.")

	flag.StringVar(&config.gmail, "gmail", "", "Gmail account. If this is present send email using the Gmail smtp server. Use -pass flag to specify the Gmail account password. If this flag is empty send email using `mail` command line program.")
	flag.StringVar(&config.pass, "pass", "", "Gmail account password. Only relevant when using -gmail flag.")
	flag.StringVar(&config.to, "to", "", "When a check fails send an email on this email address. If is empty the email alert is disabled.")

	flag.Var(&config.urls, "urls", "Comma-separated list of URLs to check.")

	version := flag.Bool("v", false, "Print version.")
	iniflags.Parse()

	if *version {
		fmt.Println("0.2.1")
		os.Exit(2)
	}

	if config.quiet {
		log.SetOutput(ioutil.Discard)
	} else if config.logFile != "" {
		file, err := os.OpenFile(config.logFile, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0644)
		if err == nil {
			log.SetOutput(file)
		}
	}

	if config.urlFile != "" {
		file, err := os.Open(config.urlFile)
		defer file.Close()
		if err != nil {
			panic("Can not open urlFile '" + config.urlFile + "'")
		}
		scanner := bufio.NewScanner(file)
		for scanner.Scan() {
			config.urls = append(config.urls, scanner.Text())
		}
		if err := scanner.Err(); err != nil {
			panic(err)
		}
	}
}

func check(client *http.Client, u string) error {
	resp, err := client.Get(u)
	if err != nil || resp.StatusCode != 200 {
		var checkError error
		if err != nil {
			checkError = err
		} else {
			checkError = &url.Error{Op: "Get", URL: u, Err: fmt.Errorf("status code is %d", resp.StatusCode)}
		}
		return checkError
	}
	return nil
}

func main() {
	var (
		config   Config
		sleep    time.Duration
		mailer   *Mailer
		client   *http.Client
		errors   []error
		errCount int
	)
	configure(&config)
	if len(config.urls) == 0 {
		log.Println("Nothing to check. Exiting...")
		return
	}
	client = &http.Client{Timeout: time.Second * time.Duration(config.timeout)}
	mailer = &Mailer{gmail: config.gmail, pass: config.pass, to: config.to}
	log.Printf("Starting to check these urls => %v...\n", config.urls)
	for {
		errors = make([]error, 0)
		for _, url := range config.urls {
			err := check(client, url)
			if err != nil {
				errors = append(errors, err)
				log.Printf("Error for '%s': %s\n", url, err.Error())
			}
		}
		errCount = len(errors)
		if errCount == 0 {
			sleep = time.Second * time.Duration(config.sleep)
		} else {
			mailer.BuildAndSendEmail(errors)
			sleep = time.Second * time.Duration(config.sleepWithError)
		}
		if config.oneCheck {
			// f(x) = [e^(-1/x^2)]
			//exitStatus := int(math.Ceil(math.Pow(2.7182818284, -1/math.Pow(float64(errCount), 2.0))))
			//os.Exit(exitStatus)
			if errCount == 0 {
				os.Exit(0)
			} else {
				os.Exit(1)
			}
		}
		time.Sleep(sleep)
	}
}
